File: 2016/python/day19/day19.py
import time
import math


def input():
    raw = 3_014_603
    # raw = 5
    return [(i + 1, 1) for i in range(raw)]


def part1():
    seats = input()

    while len(seats) > 1:
        for i in range(len(seats)):
            (elf, presents) = seats[i]
            if presents == 0:
                continue

            nextSeat = (i + 1) % len(seats)
            nextElf = seats[nextSeat]

            presents += nextElf[1]
            seats[i] = (elf, presents)
            seats[nextSeat] = (nextElf[0], 0)

        seats = [(elf, presents) for (elf, presents) in seats if presents > 0]

    return seats[0][0]


def part2():
    class Elf:
        def __init__(self, id, presents):
            self.id = id
            self.presents = presents
            self.next = None
            self.prev = None

        def __str__(self):
            return f"elf {self.id} with {self.presents} presents"

    class SeatedElves:
        def __init__(self):
            self.count = 0
            self.elves = []

        def add(self, elf):
            elfo = Elf(elf[0], elf[1])
            self.elves.append(elfo)
            elfo.next = self.elves[0]
            elfo.prev = self.elves[self.count - 1]

            self.elves[self.count - 1].next = elfo
            self.elves[0].prev = elfo

            self.count += 1

        def head(self):
            return self.elves[0]

        def remove(self, elf):
            elf.prev.next = elf.next
            elf.next.prev = elf.prev
            self.count -= 1

        def get_target(self, elf):
            target = elf
            for _ in range(math.floor(self.count / 2)):
                target = target.next
            return target

    seats = SeatedElves()
    for elf in input():
        seats.add(elf)

    elf = seats.head()
    while seats.count > 1:
        print(seats.count)
        target = seats.get_target(elf)
        elf.presents += target.presents
        seats.remove(target)
        elf = elf.next

    print(elf)


t1 = time.time()
sol = part1()
t2 = time.time()
print(f"Part 1 ({round(1000 * (t2 - t1), 3)} ms)")
print("   ", sol)
print()

t1 = time.time()
sol = part2()
t2 = time.time()
print(f"Part 2 ({round(1000 * (t2 - t1), 3)} ms)")
print("   ", sol)

File: 2016/python/day19/day19.js
class Elf {
  constructor(id) {
    this.id = id;
    this.next = null;
    this.prev = null;
  }
}

class ElfCircle {
  constructor(howManyElves) {
    this.length = howManyElves;

    this.head = new Elf(1);
    let tail = this.head;

    const halfsies = Math.floor(howManyElves / 2) + 1;

    for (let i = 2; i <= howManyElves; i += 1) {
      const elf = new Elf(i);
      elf.prev = tail;
      tail.next = elf;
      elf.next = this.head;
      tail = elf;
      if (i === halfsies) {
        this.target = elf;
      }
    }
    this.head.prev = tail

    this.player = this.head;
  }

  remove(elf) {
    if (this.head === elf) {
      this.head.prev.next = elf.next;
      this.head.next.prev = elf.prev;
      this.head = elf.next;
    }

    elf.prev.next = elf.next;
    elf.next.prev = elf.prev;
    this.length -= 1;
    if (this.length % 2 === 0) {
      this.target = this.target.next;
    } else {
      this.target = this.target.prev;
    }
  }

  next() {
    this.player = this.player.next;
    this.target = this.target.next;
  }

  toString() {
    const str = [];
    let elf = this.head;
    for (let i = 0; i < this.length; i += 1) {
      str.push(elf.id);
      elf = elf.next;
    }
    return str.join(" -> ");
  }
}

const raw = 3_014_603;
// const raw = 5;

const part1 = () => {
  const circle = new ElfCircle(raw);
  while (circle.length > 1) {
    circle.remove(circle.player.next)
    circle.next()
  }

  return circle.player.id;
};

const part2 = () => {
  const circle = new ElfCircle(raw);
  while (circle.length > 1) {
    circle.remove(circle.target);
    circle.next();
  }

  return circle.player.id;
};

const t1 = performance.now();
const sol1 = part1();
const t2 = performance.now();
console.log(`Part 1 (${Math.round((t2 - t1) * 1000) / 1000} ms)`);
console.log("  ", sol1);

const t3 = performance.now();
const sol2 = part2();
const t4 = performance.now();
console.log(`Part 2 (${Math.round((t4 - t3) * 1000) / 1000} ms)`);
console.log("  ", sol2);
